src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlHelperTest.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Microsoft.Data.SqlClient.Tests
+{
+    public class SqlHelperTest
+    {
+        private void TimeOutATask()
+        {
+            TaskCompletionSource<bool> tcs = new TaskCompletionSource<bool>();
+            AsyncHelper.WaitForCompletion(tcs.Task, 1); //Will time out as task uncompleted
+            tcs.SetException(new TimeoutException("Dummy timeout exception")); //Our task now completes with an error
+        }
+
+        private Exception UnwrapException(Exception e)
+        {
+            return e?.InnerException != null ? UnwrapException(e.InnerException) : e;
+        }
+
+        [Fact]
+        public void WaitForCompletion_DoesNotCreateUnobservedException()
```

Comment:
This test does need a bit of rework to handle randomness. It fails randomly in pipelines.

You'll notice if you run this test in loop, the second round does not pass ever. Any thoughts why can't we run this test in second round? It maybe a hint towards random errors.

Reply:
Was just looking into this - it seems the fix I had written to WaitForCompletion wasn't actually a fix. I needed to make the continuation actually observe the exception, the behaviour I had lead to a race condition, where if the continuation hadn't been executed at the point we called GC.Collect, then there was still a reference to the original task and it wasn't Collected. Creating a false positive for me having fixed it.

I proved this by modifying WaitForCompletion to return the continuation, and waited for it to complete, at which point the error happened deterministically. Making the continuation observe the exception fixed the error. My bad, I misinterpreted the text here: https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcontinuationoptions?view=netcore-3.1 saying "If you do not access the Exception property, the exception is unhandled." To only apply to OnlyOnFaulted, when it applies to all TaskContinuationOptions.

I can't see a way to make the failure deterministic without passing some sort of signal back from the continuation, which doesn't fit the signature and isn't desirable.

Reply:
This part is just a test so make it as messy as it needs to be to get a deterministic answer.

Reply:
It does pass deterministically now with the fix in place - what I couldn't get deterministic was the failure without the fix in place.